### PR TITLE
Dual Bound

### PR DESCRIPTION
**Pull request by beljadid**
## Pull request checklist
Did I miss something ?  

C++
- [x] Create dualBound.hpp 
- [x] Create  dualBound.cpp 
- [x] Add an entry in libecole/CMakeLists.txt.

C++ test :
- [x] Create test-dualbound.cpp 
- [x] Add an entry in libecole/tests/CMakeLists.txt.

Python
- [x] Enable using the new reward 
- [x] Test with python

## HelloWorld implementation

To add a helloWorld Reward (resp. observation) function :
- [ ] Create helloWorld.hpp in libecole/include/ecole/reward  (resp. libecole/include/ecole/observation)
- [ ] Create  helloWorld.cpp in libecole/src/reward  (resp. in libecole/src/observation).
- [ ] Add an entry in libecole/CMakeLists.txt.

To test  c++ helloWorld Reward (resp. observation) function :
- [ ] Create test-helloworld.cpp in libecole/tests/src/reward  (resp. libecole//tests/src/observation)
- [ ] Add an entry in libecole/tests/CMakeLists.txt.

To enable using the new reward (resp observation) with python 
- [ ] Add entries in python/src/ecole/core/reward.cpp (resp. python/src/ecole/core/observation.cpp)

#include "ecole/reward/helloWorld.hpp"
and

auto helloworld= py::class_<HelloWorld>(m, "HelloWorld", R"(		
		The reward (observation) is helloworld .
	)");
	helloworld.def(py::init<>());
	def_operators(helloworld);
	def_before_reset(helloworld, "Reset the helloworld.");
	def_extract(helloworld, R"(
		Extarct the helloworld.
		)");

To test python helloWorld Reward (resp. observation) function :
- [ ] add entries in python/tests/test_reward.py (resp. python/tests/test_observation.py)

### CODE REVIEW

File: libecole/include/ecole/reward/dualbound.hpp [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#pragma once
+
+#include "ecole/reward/abstract.hpp"
+#include "ecole/scip/type.hpp"
+
+#define infinity   1e+20
```
Comment (gasse) [inline libecole/include/ecole/reward/dualbound.hpp:6]: I think we should rather use SCIP's infinity, which should be available somewhere from SCIP

File: libecole/include/ecole/reward/dualbound.hpp [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+namespace ecole::reward {
+
+class DualBound : public RewardFunction {
+public:
+	void before_reset(scip::Model& model) override;
+	Reward extract(scip::Model& model, bool done = false) override;
+
+private:
+//	scip::real dual_bound_value = 0.0;
+	scip::real dual_bound_value = - infinity;
```
Comment (gasse) [inline libecole/include/ecole/reward/dualbound.hpp:17]: This reward function does not need to keep a memory of the past, so I think this field is not necessary.

File: libecole/src/reward/dualbound.cpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+	switch (model.get_stage()) {
+	// Only stages when the following call is authorized
+	case SCIP_STAGE_TRANSFORMED:
+  case SCIP_STAGE_INITPRESOLVE:
+  case SCIP_STAGE_PRESOLVING:
+  case SCIP_STAGE_EXITPRESOLVE:
+  case SCIP_STAGE_PRESOLVED:
+  case SCIP_STAGE_INITSOLVE:
+  case SCIP_STAGE_SOLVING:
+  case SCIP_STAGE_SOLVED:
```
Comment (gasse) [inline libecole/src/reward/dualbound.cpp:16]: Alignment should be fixed.

File: libecole/src/reward/dualbound.cpp [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+		return SCIPgetDualbound(model.get_scip_ptr());
+	default:
+		return decltype(SCIPgetDualbound(nullptr)){0};
+	}
+}
+
+void DualBound::before_reset(scip::Model& /*Model*/) {
+  	dual_bound_value = - infinity;
```
Comment (gasse) [inline libecole/src/reward/dualbound.cpp:24]: There is no need to do that.

File: libecole/tests/src/reward/test-dualbound.cpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+#include "conftest.hpp"
+#include "reward/unit-tests.hpp"
+
+using namespace ecole;
+
+TEST_CASE("DualBound unit tests", "[unit][reward]") {
+	reward::unit_tests(reward::DualBound{});
+}
+
+TEST_CASE("DualBound returns the dyal bound value", "[reward]") {
```
Comment (gasse) [inline libecole/tests/src/reward/test-dualbound.cpp:14]: Typo dyal -> dual

File: libecole/tests/src/reward/test-dualbound.cpp [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+	auto reward_func = reward::DualBound{};
+	auto model = get_model();  // a non-trivial instance is loaded
+
+	SECTION("DualBound iterations is infinity before presolving") {
+		reward_func.before_reset(model);
+//    REQUIRE(reward_func.extract(model) == Infinity);
```
Comment (gasse) [inline libecole/tests/src/reward/test-dualbound.cpp:20]: Why is this commented out ? Does this test not work ?

File: python/src/ecole/core/reward.cpp [hunk 7]
```diff
@@ -167,6 +168,19 @@ void bind_submodule(py::module_ const& m) {
...
+	auto dualbound = py::class_<DualBound>(m, "DualBound", R"(
+		dual bound.
+
+		The reward is defined as the value of dual bound.
+	)");
+	dualbound.def(py::init<>());
+	def_operators(dualbound);
+	def_before_reset(dualbound, "Reset the dual bound : set -infinity.");
+	def_extract(dualbound, R"(
+		Extarct the dual bound.
```
Comment (gasse) [inline python/src/ecole/core/reward.cpp:181]: Typo

File: libecole/tests/CMakeLists.txt [hunk 8]
```diff
@@ -34,6 +34,8 @@ add_executable(
 	src/reward/test-isdone.cpp
 	src/reward/test-nnodes.cpp
 	src/reward/test-solvingtime.cpp
+	src/reward/test-dualbound.cpp
+
```
Comment (AntoinePrv) [inline libecole/tests/CMakeLists.txt:38]: No need for new line

File: libecole/tests/src/reward/test-dualbound.cpp [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+	}
+	
+
```
Comment (AntoinePrv) [inline libecole/tests/src/reward/test-dualbound.cpp:23]: Single blank line

File: libecole/tests/src/reward/test-dualbound.cpp [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+			{"presolving/maxrounds", 0},
+			{"lp/iterlim", 0},
+			{"lp/rootiterlim", 0},
+			{"limits/totalnodes", 1},
+		});
+		advance_to_root_node(model);
+		reward_func.before_reset(model);
+//		REQUIRE(reward_func.extract(model) == Infinity);
+	}
+}
```
Comment (AntoinePrv) [inline libecole/tests/src/reward/test-dualbound.cpp:36]: No new line at end of file

File: libecole/tests/src/reward/test-dualbound.cpp [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+		model = get_model();
+		model.set_params({
+			{"presolving/maxrounds", 0},
+			{"lp/iterlim", 0},
+			{"lp/rootiterlim", 0},
+			{"limits/totalnodes", 1},
+		});
+		advance_to_root_node(model);
+		reward_func.before_reset(model);
+//		REQUIRE(reward_func.extract(model) == Infinity);
```
Comment (AntoinePrv) [inline libecole/tests/src/reward/test-dualbound.cpp:34]: Also commented out.

File: libecole/tests/src/reward/test-dualbound.cpp [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+	
+	SECTION("No dual bound value if SCIP is not solving LPs") {
+		model = get_model();
+		model.set_params({
+			{"presolving/maxrounds", 0},
+			{"lp/iterlim", 0},
+			{"lp/rootiterlim", 0},
+			{"limits/totalnodes", 1},
+		});
+		advance_to_root_node(model);
```
Comment (AntoinePrv) [inline libecole/tests/src/reward/test-dualbound.cpp:32]: Perhaps test by using the reward on function with `model` without calling `advance_to_root_node`.

File: python/src/ecole/core/reward.cpp [hunk 13]
```diff
@@ -167,6 +168,19 @@ void bind_submodule(py::module_ const& m) {
 
 		The difference in solving time is computed in between calls.
 		)");
+	
+	auto dualbound = py::class_<DualBound>(m, "DualBound", R"(
+		dual bound.
```
Comment (AntoinePrv) [inline python/src/ecole/core/reward.cpp:173]: This is not useful in the documentation. Perhaps "The problem global dual bound."
We could also add whether or not the is differential or a metric (see main comment in the PR discussion)

**Review by AntoinePrv — CHANGES_REQUESTED** [review]
Thanks @beljadid !

A lot of the formatting details can be fixed automatically using the `clang-tidy`, `clang-format`. See the [contribution doc](https://doc.ecole.ai/master/contributing.html#coding-standards) for how to enable them.

@gasse @dchetelat 
 - Is this a differential reward or just the SCIP metric. Others rewards such as `SolvingTime`, `NNodes`, `LPiterations` return the finite difference eventhough they the same name as a SCIP function.
 Switching from one to the other is easy using `.cumsum()` and `.diff()` (not yet implemented) so we should aim for consistency.
 - We should define a policy for what to return when not the SCIP is not in a valid State: `None`? `inf`? `SCIPinfinity`? `nan`?
   In this case infinity sign would depend on max/min`

**Pull request closed without merging**